api/src/controllers/sso.js

```diff
@@ -0,0 +1,200 @@
+const { createHmac } = require('crypto');
+
+const config = require('../config');
+const db = require('../db');
+const dataContext = require('../services/data-context');
+const { users } = require('@medic/user-management')(config, db, dataContext);
+const logger = require('@medic/logger');
+const secureSettings = require('@medic/settings');
+const environment = require('@medic/environment');
+const request = require('@medic/couch-request');
+
+const client = require('../openid-client-wrapper');
+const { setCookies, sendLoginErrorResponse } = require('./login');
+const settingsService = require('../services/settings');
+const { setTimeout } = require('later');
+
+const OIDC_CLIENT_SECRET_KEY = 'oidc:client-secret';
+
+const SERVER_ERROR = new Error({ status: 500, error: 'An error occurred when logging in.' });
+
+let ASConfig;
+let code_verifier;
+
+const networkCallRetry = async (call, retryCount = 3) => {
+  try {
+    return await call();
+  } catch (err) {
+    if (retryCount === 1) {
+      throw err;
+    }
+    logger.debug(`Retrying ${call.name}.`);
+    return await setTimeout(networkCallRetry(call, --retryCount), 10);
+  }
+};
+
+const init = async () => {
+  const settings = await settingsService.get();
+  if (!settings.oidc_provider) {
+    logger.info('Authorization server config settings not provided.');
+    return;
+  }
+
+  const clientSecret = await secureSettings.getCredentials(OIDC_CLIENT_SECRET_KEY);
+  if (!clientSecret) {
+    const err = `No OIDC client secret '${OIDC_CLIENT_SECRET_KEY}' configured.`;
+    logger.error(err);
+    throw new Error(err);
+  }
+
+  try {
+    ASConfig = await networkCallRetry(
+      () => client.discovery(
+        new URL(settings.oidc_provider.discovery_url), settings.oidc_provider.client_id, clientSecret
+      ),
+      3
+    );
+  } catch (e) {
+    logger.error(e);
+    const err = 'The SSO provider is unreachable.';
+    throw new Error({ status: 503, error: err });
+  }
+
+  logger.info('Authorization server config auth config loaded successfully.');
+
+  return ASConfig;
+};
+
+const getAuthorizationUrl = async (serverConfig, redirectUrl) => {
+  code_verifier = client.randomPKCECodeVerifier();
```

Comment:
We decided not to worry about the PKCE process in the MVP code. The main challenge is around handling concurrent OIDC login requests. Your current code here will not work for that since the `code_verifier` will be re-written if one login request starts before the other one finishes.  

If you check the version of the [design doc](https://docs.google.com/document/d/1LUn1ZRetAmYE04CtdcTmp-bEBvl37AZ0CvFXZChXqfU/edit?tab=t.0#heading=h.w42npptux3pi) called `PKCE Removed`, that is when I pull all the PKCE from the design.  We decided that because we were using the confidential client flow and PKCE is not _required_ by the OIDC spec (only "recommended" for confidential clients) we would not try to worry about caching code verifiers for our MVP implementation.

Reply:
I think here the issue was the identity provider enforcing PKCE but I have not revisited this since initial experiments. Looking into this.